File: Main/ArchiMeter.Analysis.FS/DependencyAnalyzer.fs
namespace ArchiMeter.Analysis
open System.Threading.Tasks
open ArchiMeter.Common

type DependencyAnalyzer() = 
    let flatten = fun r -> r
    let rec getDependencyChain(chain : DependencyChain, source : seq<EdgeItem>) = 
        match chain.IsCircular with
        | true -> [|chain|] |> Seq.ofArray
        | _ -> source 
               |> Seq.where chain.IsContinuation
               |> Seq.collect (fun i -> getDependencyChain(new DependencyChain(chain.ReferenceChain, chain.Root, i), source))
    member this.GetCircularReferences(items : seq<EdgeItem>) =
        Async.StartAsTask(
            async {
                return items 
                |> Seq.map (fun e -> async { return getDependencyChain(new DependencyChain(Seq.empty, e, e), items) } )
                |> Async.Parallel
                |> Async.RunSynchronously
                |> Seq.collect flatten
                |> Seq.where (fun (c : DependencyChain) -> c.IsCircular) 
                |> Seq.distinct
        })

File: Main/ArchiMeter.Analysis/DependencyAnalyzer.cs
namespace ArchiMeter.Analysis
{
	using System.Collections.Generic;
	using System.Linq;
	using System.Threading.Tasks;
	using Common;

	public class DependencyAnalyzer
	{
		public Task<IEnumerable<DependencyChain>> GetCircularReferences(IEnumerable<EdgeItem> items)
		{
			return Task.Factory.StartNew(() => items.SelectMany(e => GetDependencyChain(new DependencyChain(Enumerable.Empty<EdgeItem>(), e, e), items))
			                                        .Where(c => c.IsCircular)
			                                        .Distinct());
		}

		private IEnumerable<DependencyChain> GetDependencyChain(DependencyChain chain,
		                                                        IEnumerable<EdgeItem> source)
		{
			return chain.IsCircular
				       ? new[] { chain }
				       : source.Where(chain.IsContinuation).SelectMany(i => GetDependencyChain(new DependencyChain(chain.ReferenceChain, chain.Root, i), source));
		}
	}
}